File: src/LeetCode/Algorithms/Java/Single_Number.java
package LeetCode.Algorithms.Java;

import java.util.HashSet;

// https://leetcode.com/problems/single-number/
public class Single_Number {
	public static void main(String[] args) {
		int[] nums = {2, 2, 1};
		
		System.out.println(singleNumber(nums));
	}
	
	public static int singleNumber(int[] nums) {
		HashSet<Integer> hashSet = new HashSet<>();
		
		for (int i : nums) {
			if (hashSet.contains(i)) {
				hashSet.remove(i);
			} else {
				hashSet.add(i);
			}
		}
		
		return (int) hashSet.toArray()[0];
	}
}

File: src/LeetCode/Algorithms/JavaScript/Single_Number.js
/**
 * https://leetcode.com/problems/single-number/
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function(nums) {
	for (let i = 1; i < nums.length; i++) {
		nums[0] = nums[0] ^ nums[i];
	}
	
	return nums[0];
	
	/*let set = new Set();
	
	for (let n of nums) {
		if (set.has(n)) {
			set.delete(n);
		} else {
			set.add(n);
		}
	}
	
	return set.keys().next().value;*/
};

let nums = [2, 2, 1];

console.log(singleNumber(nums));